### DISCUSSION TITLE
wagmi reading struct structure of smart contract can't return key-value structure?

### DISCUSSION THREAD

**Discussion by the-content** [wevm/wagmi#2866] [General]
When using 0.x version of wagmi and ethers.js, when I read a struct in solidity, there are two types of data structures returned, one with `index:value` and one with `key:value`, now that I've upgraded wagmi, which uses viem, when I read it, it only returns `index:value`, which is leading to a problem for me. can't fetch the data better.
Contract example:
```solidity
struct myStruct {
    address player;
    uint256 amount.
}
mapping(address => myStruct ) public MyStruct ;
```

The data I want returned:

```bash
{
    0: 0x00000,
    1: 4250000000000000000000000n,
    player: 0x00000.
    amount: 4250000000000000000000000n
}
```

The data returned now:

```bash
{
    0: 0x00000,
    1: 425000000000000000000n
}
```

Did I make a writing error somewhere?
I've tried to think that it might be my abi's import issue

```js
import myAbi from "config/abi/myAbi.json" ;
```

**Comment by tmm** [comment]
What does the ABI look like? My guess is you are running into this issue: https://viem.sh/docs/faq.html#why-is-a-contract-function-return-type-returning-an-array-instead-of-an-object

**Reply by the-content** [reply]
I read the linked document you gave, and I still don't understand how to do what I mentioned in my question, how to return mixed arrays+objects like ethersjs, because, in my front-end, I need to get the result of a struct, referencing it in more than one place, such as `struct.name1` and `struct.name2` in this way, instead of Instead of `struct[0]` and `struct[1]`, the first way is more readable and maintainable, for example, if I change the order of the variables in my solidity, then the order of getting the variables in the front-end needs to be changed a lot!

**Reply by tmm** [reply]
Sorry, that document was misleading. Thought this was a different problem. The issue here is two-fold:

**Hybrid array-objects.** Ethers hybrid array-objects are not serializable so we've tried to stay away from them. We are re-evaluating this decision. If we do anything, the return type will likely be an object with index as the key for unnamed parameters. In this case, as long as the ABI has the parameter names the result will be:

```ts
{
  player: 0x00000,
  amount: 4250000000000000000000000n,
}
```

If one of the parameters is not named, let's say `'player'`, the result will look like this:

```ts
{
  0: 0x00000,
  amount: 4250000000000000000000000n,
}
```

This will all be type-safe so as long [as you set up your ABIs correctly](https://wagmi.sh/react/typescript#type-inference), you will have nice autocomplete and safe access.

**ABI output lack of specificity.** The type system internal to wagmi/viem is ABI-driven. You might think you are returning a struct on the contract-level, but be confused why it's not reflected by wagmi/viem. Usually, the answer is the ABI specifies an array, not a tuple type corresponding to the struct For example, given the following contract:

```sol
contract Example {
  struct Foo {
    uint256 a;
    bool b;
  }

  Foo public foo;
  function getFoo() public returns (Foo)
  function getArray() public returns (uint256 a, bool b)
}
```

The `Foo public foo` ABI definition looks like:

```json
{
  "inputs": [],
  "name": "foo",
  "outputs": [
    {
      "internalType": "uint256",
      "name": "a",
      "type": "uint256"
    },
    {
      "internalType": "bool",
      "name": "b",
      "type": "bool"
    }
  ],
  "stateMutability": "view",
  "type": "function"
}
```

Which ends up the same as `getArray`. While `getFoo`'s ABI definition looks like:

```json
{
  "inputs": [],
  "name": "getFoo",
  "outputs": [
    {
      "components": [
        {
          "internalType": "uint256",
          "name": "a",
          "type": "uint256"
        },
        {
          "internalType": "bool",
          "name": "b",
          "type": "bool"
        }
      ],
      "internalType": "struct Example.Foo",
      "name": "",
      "type": "tuple"
    }
  ],
  "stateMutability": "nonpayable",
  "type": "function"
}
```

Since `getFoo` outputs a `'tuple'`, we convert it into an object in wagmi/viem whereas `foo` (a global public value) does not preserve this info in the ABI. My guess is that this is happening to you (never shared you ABI so can't confirm though).

Overall, we are trying to stay as close to the platform as possible and not inject additional magic/opinions into how we represent types. This allows for things to be more resilient to possible ABI schema changes in the future and reduces runtime code.

**Reply by the-content** [reply]
This is the struct in my contract:

```
struct Profile {
  NodeType nodeType;
  address inviter;
  uint256 firstInvestTime; 
  uint256 validDirectInviteCount;
  uint256 totalInviteCount;
  bool isRemoveBlackListFlag;
}

mapping(address => Profile) public Profiles;
```

Below is the abi data about Profiles I compiled in remix: 

```
[
  {
    "inputs": [
      {
        "internalType": "address", 
        "name": "",
        "type": "address"
      }
    ],
    "name": "Profiles", 
    "outputs": [
      {
        "internalType": "enum PEPEDIFI.NodeType",
        "name": "nodeType", 
        "type": "uint8"
      },
      {
        "internalType": "address",
        "name": "inviter",
        "type": "address"
      },
      {
        "internalType": "uint256",
        "name": "firstInvestTime",
        "type": "uint256"
      },
      {
        "internalType": "uint256",
        "name": "validDirectInviteCount",
        "type": "uint256"
      },
      {  
        "internalType": "uint256",
        "name": "totalInviteCount",
        "type": "uint256"
      },
      {
        "internalType": "bool",
        "name": "isRemoveBlackListFlag", 
        "type": "bool"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  }
]
```

Below is the profiles data I got in console:

```
[0, '0x0000000000000000000000000000000000000000', 0n, 0n, 0n, false]
```

I imported the compiled abi file from remix in javascript:

```
import myAbi from "config/abi/myAbi.json" 
```

Then I requested in component:

```
const {
  data: profile, 
} = useContractRead({
  address: CONTRACTS.PEPEDEFI,
  abi: myAbi,
  functionName: "Profiles",
  args: [address],
  enabled: Boolean(address)
})
```

The `profile` data I got is:

```
[0, '0x0000000000000000000000000000000000000000', 0n, 0n, 0n, false]
``` 

What I expect is:

```
{
  0: 0,
  1: '0x0000000000000000000000000000000000000000', 
  2: 0n,
  3: 0n,
  4: 0n,
  5: false,
  nodeType: 0,
  inviter: '0x0000000000000000000000000000000000000000',
  firstInvestTime: 0n,
  validDirectInviteCount: 0n,
  totalInviteCount: 0n,
  isRemoveBlackListFlag: false
}
```

I directly imported the `myAbi.json` file without any processing. I didn't need to do any processing before using previous versions of wagmi, and could directly get the data I wanted. But now I need to do some processing, but I don't know what to do. Please help me, thank you!

**Reply by tmm** [reply]
Please read my previous message, it explains everything in detail. This [TypeScript Playground](https://www.typescriptlang.org/play?ts=5.0.4#code/JYWwDg9gTgLgBAbzgVwM4FMDCEB2MoCGAxjAEroEAmcAvnAGZQQhwDkA7gQOYjCsBQ-IrlTwCAI2ABGOAF44AbX5xEylXGA4wyGKgBcipJpjooOAgBsAKgE8w6A6yqUo6VKlYAaODDsO2zq7urLQAup5qKuYg-qwACkz0wBZuXpFwEDraugZK6upI0bE4EJTotvZePn6OyMYAHCE0EfkqhQQxjpoAbsAmUFW+9o6Bbh60La3tnWxJUKIAkjjdbjBWoOiDNWx1eABMAKwAbE2T+dOx3ZbAlAAiwK4kS70m2Mh4W8M7xocnE+ltOBFRwwCAwSzPPpYTIfbxDWK7GC-U4AxBAjqxYCocggCArABCFmIAGsADJYmAAMSJXE+sXEEAgFhR+XC6VEBBMAFkdBJkn0bI5euh2Gl8vDHPR3iRgLgxf9QnACKg4MIcKJBGrRGjKJyCAYwIlkugZHR5GhoXhCCRyFQABQINSjdyOAAMAA95XyDHypJMpTgZbgAHIYxwJCBJFIeSYEKBcfSKVge1hsmgASn4hsjxqk-AA9Pm4AA9AD8mpEYkkezkijUjvymmyiYURjwpnM1m2TkoLjGdJGvaC4xobPywLYEajqTOGSyOhbqIbrSiYcnRpS8pXErYMGQYE3s-ywnAuHQeEXK-O6JmrBKZQqmzh3cRjX+V4KN8xyyhA2fXx7PtgnfD80QnVg5kWZZVnWGIB2+fZjhZUCLiFa47gedAnh-V4YRgeDWERZEQI-VDdzBCEcOhd58P-BEfiQkirzI1gsRxPF0EJElyVEaluAIhkmWQlcx1aZp0lElQOW5XlJAsAUhWAEUt2qACAyDHB5XExVlVVSsK3VeAkF1cEDQ3dAazNFAMGwK1iDIChKAdJ0hzGN1PVjSQfWrf1pRgWUcFDW8p2NGMnXjFtk09NNM2zac9gLIsy34IA) also helps explain.

**Reply by jaybuidl** [reply]
Confirmed it works by returning a struct in the contract. But modifying your contracts for the sake of getting named parameters in your frontend (and avoid likely errors) is a luxury that many people don't have, e.g. if the contract has been audited or has specific gas optimization constraints. 

I'd recommend exploring a solution which doesn't require changing the contracts.

**Reply by tmm** [reply]
The ABI is the source of truth. Better that compilers output ABIs like this upstream than manually editing or Wagmi (Viem) introducing a new abstraction.
